Fixed #74862: Unable to clone instance when private __clone defined

DanielCiochiu opened the pull request:
Even though __clone was implemented as private and called only from parent class, child extending class instance could not be cloned.

nikic commented on Zend/tests/generators/bug74862.phpt:
```diff
@@ -0,0 +1,43 @@
+?>
+--EXPECT--
+object(c)#2 (0) {
+}
```
This test should go into Zend/tests rather than Zend/tests/generators, as it's not generator related.

nikic commented on Zend/zend_vm_execute.h:
```diff
@@ -48790,7 +48790,7 @@ static ZEND_OPCODE_HANDLER_RET ZEND_FASTCALL ZEND_CLONE_SPEC_TMPVAR_HANDLER(ZEND
 			 */
 			scope = EX(func)->op_array.scope;
-			if (UNEXPECTED(ce != scope)) {
+                if (UNEXPECTED(clone->common.scope != scope)) {
```
zend_vm_execute.h is a generated file. The change should be made in zend_vm_def.h and then zend_vm_execute.h can be regenerated using zend_vm_gen.php.

nikic replied:
I believe this will also not work if the extending class also defines a private `__clone()`. If it doesn't, please add a test and use the `zend_check_private` function.

DanielCiochiu commented:
Sorry about above. Moved the change inside zend_vm_def.h. and updated the condition to use zend_check_private as mentioned above (clone implemented in child also didn't work).
Moved test file from generators to tests and added another test file for the child clone case.

DanielCiochiu commented:
Not sure if generated files also need to be pushed.

nikic commented on Zend/zend_vm_def.h:
```diff
@@ -4994,8 +4994,8 @@ ZEND_VM_HANDLER(110, ZEND_CLONE, CONST|TMPVAR|UNUSED|THIS|CV, ANY)
-			if (UNEXPECTED(ce != scope)) {
-				zend_throw_error(NULL, "Call to private %s::__clone() from context '%s'", ZSTR_VAL(ce->name), scope ? ZSTR_VAL(scope->name) : "");
+			if (!zend_check_private(clone, scope, clone->common.function_name)) {
+				zend_throw_error(NULL, "Call to private %s::__clone() from context '%s'", ZSTR_VAL(scope->name), scope ? ZSTR_VAL(scope->name) : "");
```
I believe the error message should be using `clone->common.scope` for the first part.

nikic commented:
@DanielCiochiu Yep, generated files should be pushed as well, otherwise CI will fail.

DanielCiochiu commented:
Changed error message to display the correct scope in which the called method is implemented. Added generated zend_vm_execute.h file. Not sure why builds fail.

tpunt commented:
The test `tests/classes/factory_and_singleton_008.phpt` is failing with a segfault. `scope` can be `NULL` when `__clone` is not invoked in a class context (it should simply emit an `Error` exception).

DanielCiochiu commented:
Fixed, but travis fails when testing "ext/mysqli/tests/mysqli_stmt_bind_result_format.phpt" with ENABLE_MAINTAINER_ZTS=1.

tpunt commented:
The failed test does not seem related.

Regarding the latest change, the additional check should not be needed. The definition location of `__clone` will always have a scope, and so `clone->common.scope->name` should never be `NULL`.

Other than that, LGTM.

DanielCiochiu commented:
@tpunt Removed unnecessarily check for NULL scope and builds passed.

nikic commented:
Merged as https://github.com/php/php-src/commit/dba5a798a28482ce4a6475cc4de1586f11126175 into 7.1+. Thanks for your work on this issue!
